{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import networkx as nx\n",
    "from sklearn.metrics import pairwise as pw\n",
    "pd.set_option('mode.chained_assignment',None)\n",
    "import panel as pn\n",
    "import hvplot.pandas\n",
    "import hvplot.networkx as hvnx\n",
    "import holoviews as hv\n",
    "from holoviews import opts\n",
    "import matplotlib.pyplot as plt\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"PredictorData.txt\", sep='\\t', usecols=range(4,31))\n",
    "add_cols = pd.read_csv(\"PredictorData.txt\", sep='\\t', usecols=[\"Delta F3\", \"Min correl\", \"id\"])\n",
    "df = df.merge(add_cols, on=\"id\")\n",
    "df.index = df[\"id\"]\n",
    "\n",
    "df_core = df.iloc[:, 18:33]\n",
    "meta_dict = {}\n",
    "meta_dict[\"gene_id\"] = {}\n",
    "for g, i in zip(df[\"Lead gene name\"], df.index):\n",
    "    if g not in meta_dict[\"gene_id\"].keys():\n",
    "        meta_dict[\"gene_id\"][g] = i\n",
    "    else:\n",
    "        if type(meta_dict[\"gene_id\"][g]) == list:\n",
    "            meta_dict[\"gene_id\"][g].append(i)\n",
    "        else:\n",
    "            meta_dict[\"gene_id\"][g] = [meta_dict[\"gene_id\"][g], i]\n",
    "meta_dict[\"id_meta\"] = df.iloc[:, 1:18].to_dict(\"index\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_pca = pd.read_csv(\"PCA cooridnates Jurkat Exosomes 13 April 2020t.txt\", sep='\\t')\n",
    "#df_pca.columns\n",
    "#df_pca.columns = ['Primary ID', 'ClassID', 'Lead gene name',\n",
    "#       'Protein names', 'Gene names',\n",
    "#       'Fasta headers', 'Sec. ID',\n",
    "#       'Organelle', 'Min Ratio count', 'Class',\n",
    "#       'PCA 1', 'PCA2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Backend functions\n",
    "\n",
    "## single query dataset\n",
    "def hv_query_df(df_core, meta_dict, gene, dist=\"Manhattan\",\n",
    "                min_enr=-1, min_corr=0, size=50, rank_tolerance=25, perc_area=250):\n",
    "    size += 1\n",
    "    if gene.upper() not in meta_dict[\"gene_id\"].keys():\n",
    "        return \"Gene {} not in dataset\".format(gene)\n",
    "    \n",
    "    # Filter data\n",
    "    df = df_core.loc[[enr >= min_enr and corr >= min_corr for enr, corr\n",
    "                      in zip(df_core[\"Delta F3\"], df_core[\"Min correl\"])]].copy()\n",
    "    df.drop([\"Delta F3\", \"Min correl\"], axis=1, inplace=True)\n",
    "    \n",
    "    # Retrieve query\n",
    "    query_id = meta_dict[\"gene_id\"][gene]\n",
    "    msg_append = \"\"\n",
    "    if type(query_id) == list:\n",
    "        counts = [meta_dict[\"id_meta\"][qid1][\"Min Ratio count\"] for qid1 in query_id]\n",
    "        max_count = [qid2 for qid2, count in zip(query_id, counts) if count == max(counts)][0]\n",
    "        msg_append = \"Multiple ids found for gene {}: {}. Using id with the highest number of counts: {}\".format(\n",
    "            gene, str(query_id), max_count)\n",
    "        query_id = max_count\n",
    "    if query_id not in df.index:\n",
    "        return \"The query protein {} was filtered out.\".format(gene)\n",
    "    query_profile = df.loc[query_id,:]\n",
    "    \n",
    "    # Calculate difference profiles and sum within replicates\n",
    "    df_diff = df-query_profile\n",
    "    df_diff = df_diff.transform(np.abs)\n",
    "    df_diff_rep = df_diff.apply(lambda x: pd.Series([sum(x[i:i+3]) for i in [0,3,6]]), axis=1)\n",
    "    df_diff_rep.columns=[\"Replicate 1\", \"Replicate 2\", \"Replicate 3\"]\n",
    "    \n",
    "    # Score distances\n",
    "    if dist == \"Manhattan\":\n",
    "        dists = df_diff.apply(sum, axis=1)\n",
    "    dists.name = \"Distance measure\"\n",
    "    dists = dists.sort_values()\n",
    "    \n",
    "    # Calculate rank ranges across replicates\n",
    "    df_diff_rep_rank = df_diff_rep.rank()\n",
    "    df_diff_rep_rank.columns = [el + \" rank\" for el in df_diff_rep_rank.columns]\n",
    "    df_diff_rep_rank[\"Rank range\"] = df_diff_rep_rank.apply(lambda x: max(x)-min(x), axis=1)\n",
    "    df_diff_rep_rank[\"Common lists\"] = df_diff_rep_rank[[\"Replicate 1 rank\", \"Replicate 2 rank\", \"Replicate 3 rank\"]].apply(\n",
    "        lambda x: sum([el <= size+rank_tolerance for el in x]), axis=1)\n",
    "    \n",
    "    # Calculate percentiles in local neighborhood\n",
    "    area_members = dists.index[0:perc_area]\n",
    "    area_members = df_core.loc[area_members,[col for col in df_core.columns if col not in [\"Delta F3\", \"Min correl\"]]]\n",
    "    area_dists = pw.distance.pdist(np.array(area_members), \"cityblock\")\n",
    "    qsteps = [0,0.001,0.005,0.01,0.025,0.05,0.1,0.25,0.5,1]\n",
    "    q = pd.Series(area_dists).quantile(q=qsteps)\n",
    "    \n",
    "    # Calculate z-scoring of distances to query\n",
    "    z_dists_in = dists[1:perc_area]\n",
    "    median = np.median(z_dists_in)\n",
    "    dists_from_median = np.abs(z_dists_in-median)\n",
    "    mad = np.median(dists_from_median)\n",
    "    rob_sd = 1.4826*mad\n",
    "    z_dists = -(z_dists_in-median)/rob_sd\n",
    "    z_dists = pd.DataFrame(z_dists)\n",
    "    z_dists.columns = [\"z-scored closeness\"]\n",
    "    z_dists[\"z-scoring based category\"] = pd.cut(z_dists[\"z-scored closeness\"],\n",
    "                                          bins=[-1000, 1.28, 1.64, 2.33, 3.09, 1000],\n",
    "                                          labels=[\"-\", \"B\", \"*\", \"**\", \"***\"])\n",
    "    \n",
    "    #assemble output table\n",
    "    out_ids = dists.index[0:size]\n",
    "    out = pd.DataFrame(dists[out_ids])\n",
    "    out = out.join(df_diff_rep, how=\"left\")\n",
    "    out = out.join(df_diff_rep_rank, how=\"left\")\n",
    "    out = out.join(pd.DataFrame.from_dict({el:meta_dict[\"id_meta\"][el] for el in out_ids}, orient=\"index\"), how=\"left\")\n",
    "    out = out.join(df_core, how=\"left\")\n",
    "    out[\"Local distance percentile\"] = pd.cut(out[\"Distance measure\"], bins=[-1]+list(q.values)[1:],\n",
    "                                              labels=qsteps[1:])\n",
    "    out = out.join(z_dists, how=\"left\")\n",
    "    out.reset_index(inplace=True)\n",
    "    \n",
    "    msg = \"Neighborhood recalculated for {}.\\n\".format(gene)+msg_append\n",
    "    \n",
    "    return out, q, msg\n",
    "\n",
    "## query multiple genes\n",
    "def multi_query(df_core, meta_dict, genes, dist=\"Manhattan\",\n",
    "                min_enr=-1, min_corr=0, size=50, rank_tolerance=25, perc_area=250, pb=None):\n",
    "    hits = []\n",
    "    qs = []\n",
    "    msgs = \"\"\n",
    "    for gene in genes:\n",
    "        output = hv_query_df(df_core, meta_dict, gene=gene, size=size, rank_tolerance=rank_tolerance,\n",
    "                             min_enr=min_enr, min_corr=min_corr, perc_area=perc_area)\n",
    "        if len(output) == 3:\n",
    "            (h, q, m) = output\n",
    "            h[\"Query\"] = np.repeat(gene, len(h))\n",
    "            hits.append(h)\n",
    "            qs.append(q)\n",
    "        else:\n",
    "            m = output\n",
    "        msgs= msgs+m+\"\\n\"\n",
    "        if pb:\n",
    "            pb.value = pb.value+1\n",
    "    \n",
    "    query_result = pd.concat(hits)\n",
    "    q = pd.DataFrame(qs).apply(np.mean, axis=0)\n",
    "    \n",
    "    return query_result.reset_index(drop=True), q, msgs\n",
    "\n",
    "## barplot function single query\n",
    "def sq_barplot(df, var):\n",
    "    q_m = df[[\"Distance measure\", \"Rank range\", \"Delta F3\", \"Min correl\", \"Local distance percentile\",\n",
    "              \"Lead gene name\", \"Common lists\", \"z-scoring based category\"]]\n",
    "    q_m[\"Lead gene name\"] = [str(n)+\", rank \"+str(r) for n, r in zip(q_m[\"Lead gene name\"], q_m.index)]\n",
    "    label_var_dict = {\n",
    "        \"Distance\": \"Distance measure\",\n",
    "        \"z-scoring based category\": \"z-scoring based category\",\n",
    "        \"Rank range between replicates\": \"Rank range\",\n",
    "        \"Percentile change F3 - full proteome\": \"Delta F3\",\n",
    "        \"Worst replicate correlation\": \"Min correl\",\n",
    "        \"Local distance percentile\": \"Local distance percentile\"\n",
    "    }\n",
    "    q_m = q_m.iloc[::-1,:]\n",
    "    var = label_var_dict[var]\n",
    "    if var == \"z-scoring based category\":\n",
    "        z_dict = {\"-\":5, \"B\":4, \"*\":3, \"**\":2, \"***\":1, np.nan:0, None:0}\n",
    "        q_m[\"z-scoring based category\"] = [z_dict[k] for k in q_m[\"z-scoring based category\"]]\n",
    "        z_dict = {\"-\":5, \"B\":4, \"*\":3, \"**\":2, \"***\":1, \"query\":0}\n",
    "        p = q_m.hvplot.barh(x=\"Lead gene name\", y=var, color=\"Common lists\",\n",
    "                               height=60+11*len(q_m), xticks = [(z_dict[k], k) for k in z_dict.keys()])\n",
    "    else:\n",
    "        p = q_m.hvplot.barh(x=\"Lead gene name\", y=var, color=\"Common lists\",\n",
    "                               height=60+11*len(q_m))\n",
    "    p.opts(cmap=[\"orange\", \"darkgrey\", \"white\", \"lightskyblue\"][0:len(np.unique(q_m[\"Common lists\"]))][::-1])\n",
    "    p.opts(legend_position=\"top\")\n",
    "    return p\n",
    "\n",
    "## Network construction functions\n",
    "def filter_nwk_members(query_result, min_rep, max_z):\n",
    "    nwk_members_1 = query_result[query_result[\"Common lists\"] >= min_rep][\"id\"].values\n",
    "    z_dict = {\"-\":5, \"B\":4, \"*\":3, \"**\":2, \"***\":1, np.nan:0, None:0}\n",
    "    query_result[\"z-scoring based category\"] = [z_dict[k] for k in query_result[\"z-scoring based category\"]]\n",
    "    nwk_members_2 = query_result[query_result[\"z-scoring based category\"] <= max_z][\"id\"].values\n",
    "    nwk_members = [el for el in nwk_members_1 if el in nwk_members_2]\n",
    "    nwk_members = np.unique(nwk_members)\n",
    "    if len(nwk_members) == 1:\n",
    "        return None\n",
    "    else:\n",
    "        return nwk_members\n",
    "\n",
    "def layout_network(df_core, query_result, nwk_members, max_q, q, layout_method=\"Spring\"):\n",
    "    nwk_profiles = df_core.loc[nwk_members,[col for col in df_core.columns if col not in [\"Delta F3\", \"Min correl\"]]]\n",
    "    dists = pw.distance.pdist(np.array(nwk_profiles), \"cityblock\")\n",
    "    dists_pd = pd.DataFrame([[meta_dict[\"id_meta\"][nwk_members[i1]][\"Lead gene name\"],\n",
    "                              meta_dict[\"id_meta\"][nwk_members[i2]][\"Lead gene name\"]]\n",
    "                             for i1 in range(len(nwk_members)) for i2 in range(len(nwk_members)) if i1<i2])\n",
    "    dists_pd.columns = [\"Gene 1\", \"Gene 2\"]\n",
    "    dists_pd[\"dist\"] = [1/el for el in dists]\n",
    "    dists_pd[\"quantile\"] = pd.cut(pd.Series(dists), bins=[-1]+list(q.values)[1:], labels=[float(el) for el in q.index[1:]])\n",
    "    dists_pd = dists_pd[dists_pd[\"quantile\"] <= float(max_q)]\n",
    "    if len(dists_pd) == 0:\n",
    "        return \"No data at this quantile.\"\n",
    "    \n",
    "    # Genrate network graph\n",
    "    G = nx.Graph()\n",
    "    for idx in range(len(dists_pd)):\n",
    "        G.add_edge(dists_pd.iloc[idx,0], dists_pd.iloc[idx,1], weight=dists_pd.iloc[idx,2],\n",
    "                   quantile=dists_pd.iloc[idx,3])\n",
    "    if layout_method == \"Spectral\":\n",
    "        GP = nx.spectral_layout(G)\n",
    "    elif layout_method == \"Spring\":\n",
    "        GP = nx.spring_layout(G, k=5/np.sqrt(len(dists_pd)))\n",
    "    \n",
    "    return dists_pd, GP\n",
    "\n",
    "## Functions for drawing a network\n",
    "def draw_network_figure(dists_pd, GP, query_result, highlight, q):\n",
    "    # Regerate network graph from distances\n",
    "    G = nx.Graph()\n",
    "    for idx in range(len(dists_pd)):\n",
    "        G.add_edge(dists_pd.iloc[idx,0], dists_pd.iloc[idx,1], weight=dists_pd.iloc[idx,2],\n",
    "                   quantile=dists_pd.iloc[idx,3])\n",
    "\n",
    "    fig = plt.figure(figsize=(7,7))\n",
    "    fig.set_dpi(300)\n",
    "\n",
    "    # edges\n",
    "    edge_styles = [{\"width\": 5, \"edge_color\": \"darkred\", \"style\": \"solid\", \"alpha\": 1}, # 0.1%\n",
    "                   {\"width\": 5, \"edge_color\": \"darkred\", \"style\": \"solid\", \"alpha\": 1}, # 0.5%\n",
    "                   {\"width\": 5, \"edge_color\": \"darkred\", \"style\": \"solid\", \"alpha\": 1}, # 1%\n",
    "                   {\"width\": 4, \"edge_color\": \"red\", \"style\": \"solid\", \"alpha\": 1}, # 2.5%\n",
    "                   {\"width\": 4, \"edge_color\": \"red\", \"style\": \"solid\", \"alpha\": 1}, # 5%\n",
    "                   {\"width\": 3, \"edge_color\": \"darkorange\", \"style\": \"solid\", \"alpha\": 1}, # 10%\n",
    "                   {\"width\": 2, \"edge_color\": \"#ababab\", \"style\": \"solid\", \"alpha\": 1}, # 25%\n",
    "                   {\"width\": 2, \"edge_color\": \"lightgrey\", \"style\": \"solid\", \"alpha\": 1}, # 50%\n",
    "                   {\"width\": 2, \"edge_color\": \"lightgrey\", \"style\": \"dotted\", \"alpha\": 1}] # 100%\n",
    "    edges = []\n",
    "    for q_cat, style in zip(q.index[-2::-1], edge_styles[-2::-1]):\n",
    "        edgelist = [(u, v) for (u, v, d) in G.edges(data=True) if d['quantile']==float(q_cat)]\n",
    "        if len(edgelist) == 0:\n",
    "            continue\n",
    "        Gi = nx.Graph(edgelist)\n",
    "        nx.draw_networkx_edges(Gi, GP, **style)\n",
    "    \n",
    "    # nodes\n",
    "    # highlighted gene\n",
    "    if highlight is not None:\n",
    "        Gi = nx.Graph([(hit,hit) for hit in [highlight] if hit in G.nodes()])\n",
    "        nx.draw_networkx_nodes(Gi, GP, node_color=\"blue\", node_size=800)\n",
    "    # replicate 1\n",
    "    Gi = nx.Graph([(hit,hit) for hit, rep in zip(query_result[\"Lead gene name\"], query_result[\"Common lists\"])\n",
    "                   if rep==1 and hit in G.nodes()])\n",
    "    nx.draw_networkx_nodes(Gi, GP, node_color=\"white\", node_size=600)\n",
    "    nx.draw_networkx_labels(Gi, GP, font_size=7)\n",
    "    # replicate 2\n",
    "    Gi = nx.Graph([(hit,hit) for hit, rep in zip(query_result[\"Lead gene name\"], query_result[\"Common lists\"])\n",
    "                   if rep==2 and hit in G.nodes()])\n",
    "    nx.draw_networkx_nodes(Gi, GP, node_color=\"lightgrey\", node_size=600)\n",
    "    nx.draw_networkx_labels(Gi, GP, font_size=7)\n",
    "    # replicate 3\n",
    "    Gi = nx.Graph([(hit,hit) for hit, rep in zip(query_result[\"Lead gene name\"], query_result[\"Common lists\"])\n",
    "                   if rep==3 and hit in G.nodes()])\n",
    "    nx.draw_networkx_nodes(Gi, GP, node_color=\"orange\", node_size=600)\n",
    "    nx.draw_networkx_labels(Gi, GP, font_size=7)\n",
    "    # query\n",
    "    if \"Query\" in query_result.columns:\n",
    "        Gi = nx.Graph([(hit,hit) for hit in np.unique(query_result[\"Query\"]) if hit in G.nodes()])\n",
    "        nx.draw_networkx_nodes(Gi, GP, node_color=\"red\", node_size=600)\n",
    "    else:\n",
    "        Gi = nx.Graph([(hit,hit) for hit in [query_result[\"Lead gene name\"][0]] if hit in G.nodes()])\n",
    "        nx.draw_networkx_nodes(Gi, GP, node_color=\"red\", node_size=600)\n",
    "    plt.close()\n",
    "\n",
    "    return fig\n",
    "\n",
    "def draw_network_interactive(dists_pd, GP, query_result, highlight, q):\n",
    "    G = nx.Graph()\n",
    "    for idx in range(len(dists_pd)):\n",
    "        G.add_edge(dists_pd.iloc[idx,0], dists_pd.iloc[idx,1], weight=dists_pd.iloc[idx,2],\n",
    "                   quantile=dists_pd.iloc[idx,3])\n",
    "    \n",
    "    width=min(900, int(np.sqrt(len(G.nodes)))*150)\n",
    "    if width < 400:\n",
    "        width=400\n",
    "    \n",
    "    Gi = nx.Graph([(hit,hit) for hit in [highlight] if hit in G.nodes()])\n",
    "    highlight = hvnx.draw(Gi, GP, node_color=\"blue\", node_size=3000, with_labels=False,\n",
    "                        width=width, height=width)\n",
    "    Gi = nx.Graph([(hit,hit) for hit, rep in zip(query_result[\"Lead gene name\"], query_result[\"Common lists\"]) if rep==1 and hit in G.nodes()])\n",
    "    labels0 = hvnx.draw(Gi, GP, node_color=\"white\", node_size=2500, with_labels=True,\n",
    "                        width=width, height=width)\n",
    "    Gi = nx.Graph([(hit,hit) for hit, rep in zip(query_result[\"Lead gene name\"], query_result[\"Common lists\"]) if rep==2 and hit in G.nodes()])\n",
    "    labels1 = hvnx.draw(Gi, GP, node_color=\"lightgrey\", node_size=2500, with_labels=True,\n",
    "                        width=width, height=width)\n",
    "    Gi = nx.Graph([(hit,hit) for hit, rep in zip(query_result[\"Lead gene name\"], query_result[\"Common lists\"]) if rep==3 and hit in G.nodes()])\n",
    "    labels2 = hvnx.draw(Gi, GP, node_color=\"orange\", node_size=2500, with_labels=True,\n",
    "                        width=width, height=width)\n",
    "    if \"Query\" in query_result.columns:\n",
    "        Gi = nx.Graph([(hit,hit) for hit in np.unique(query_result[\"Query\"]) if hit in G.nodes()])\n",
    "        labels3 = hvnx.draw(Gi, GP, node_color=\"red\", node_size=2500, with_labels=True,\n",
    "                            width=width, height=width)\n",
    "    else:\n",
    "        Gi = nx.Graph([(hit,hit) for hit in [query_result[\"Lead gene name\"][0]] if hit in G.nodes()])\n",
    "        labels3 = hvnx.draw(Gi, GP, node_color=\"red\", node_size=2500, with_labels=True,\n",
    "                            width=width, height=width)\n",
    "    # .opts(tools=[HoverTool(tooltips=[('index', '@index_hover')])])\n",
    "\n",
    "    # edges\n",
    "    edge_styles = [{\"edge_width\": 8, \"edge_color\": \"darkred\", \"style\": \"solid\", \"alpha\": 1}, # 0.1%\n",
    "                   {\"edge_width\": 8, \"edge_color\": \"darkred\", \"style\": \"solid\", \"alpha\": 1}, # 0.5%\n",
    "                   {\"edge_width\": 8, \"edge_color\": \"darkred\", \"style\": \"solid\", \"alpha\": 1}, # 1%\n",
    "                   {\"edge_width\": 6, \"edge_color\": \"red\", \"style\": \"solid\", \"alpha\": 1}, # 2.5%\n",
    "                   {\"edge_width\": 6, \"edge_color\": \"red\", \"style\": \"solid\", \"alpha\": 1}, # 5%\n",
    "                   {\"edge_width\": 4, \"edge_color\": \"darkorange\", \"style\": \"solid\", \"alpha\": 1}, # 10%\n",
    "                   {\"edge_width\": 3, \"edge_color\": \"#ababab\", \"style\": \"solid\", \"alpha\": 1}, # 25%\n",
    "                   {\"edge_width\": 2, \"edge_color\": \"lightgrey\", \"style\": \"solid\", \"alpha\": 1}, # 50%\n",
    "                   {\"edge_width\": 2, \"edge_color\": \"lightgrey\", \"style\": \"dotted\", \"alpha\": 1}] # 100%\n",
    "    edges = []\n",
    "    for q_cat, style in zip(q.index[1:], edge_styles):\n",
    "        edgelist = [(u, v) for (u, v, d) in G.edges(data=True) if d['quantile']==float(q_cat)]\n",
    "        if len(edgelist) == 0:\n",
    "            continue\n",
    "        Gi = nx.Graph(edgelist)\n",
    "        edge = hvnx.draw(Gi, GP, node_size=2000, with_labels=False,\n",
    "                         width=width, height=width, **style)\n",
    "        edges.append(edge)\n",
    "    \n",
    "    nw = hv.Overlay(edges[::-1]) * highlight * labels0 * labels1 * labels2 * labels3\n",
    "    return nw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "## Interface elements for single query output\n",
    "\n",
    "# Panes to cache query calculation as json\n",
    "cache_sq_neighborhood = pn.pane.Str(\"\")  # neighborhood query result\n",
    "cache_sq_q = pn.pane.Str(\"\")  # neighborhood quantile steps\n",
    "cache_sq_nwdists = pn.pane.Str(\"\")  # network edge distances\n",
    "cache_sq_gp = pn.pane.Str(\"\")  # network layout positions\n",
    "\n",
    "# Main options\n",
    "input_sq_gene = pn.widgets.AutocompleteInput(options=[el for el in meta_dict[\"gene_id\"].keys() if type(el) == str],\n",
    "                                             name=\"Select gene (capital letters):\", value=\"CD63\")\n",
    "input_sq_topn = pn.widgets.IntSlider(start=20, end=100, step=5, value=50, value_throttled=50,\n",
    "                                     name=\"Neighborhood size\")\n",
    "input_sq_button = pn.widgets.Button(name='Calculate neighborhood', button_type='primary')\n",
    "output_sq_status = pn.pane.Markdown(\"\", width_policy=\"max\", sizing_mode=\"stretch_width\")\n",
    "\n",
    "# Advanced options\n",
    "input_sq_tolerance = pn.widgets.IntSlider(start=5, end=50, step=5, value=25, value_throttled=25,\n",
    "                                          name=\"Replicate neighborhood tolerance\")\n",
    "input_sq_minr = pn.widgets.FloatSlider(start=-1, end=1, step=0.05, value=0, value_throttled=0,\n",
    "                                       name=\"Minimal profile correlation\")\n",
    "input_sq_minenr = pn.widgets.FloatSlider(start=-1, end=1, step=0.05, value=-1, value_throttled=-1,\n",
    "                                         name=\"Minimal percentile shift in F3 vs. full proteome\")\n",
    "input_sq_nhsize = pn.widgets.IntSlider(start=50, end=500, step=50, value=250, value_throttled=250,\n",
    "                                       name=\"Neighborhood size for scoring\")\n",
    "\n",
    "# Tabcolumn containing options\n",
    "options_single = pn.Tabs()\n",
    "options_single.append((\"Neighborhood selection\", pn.Column(input_sq_gene, input_sq_topn,\n",
    "                                                           input_sq_button, output_sq_status)))\n",
    "options_single.append((\"Advanced options\", pn.Column(input_sq_tolerance, input_sq_minr,\n",
    "                                                     input_sq_minenr, input_sq_nhsize)))\n",
    "\n",
    "# Options for network display\n",
    "input_sq_minz = pn.widgets.Select(options=[\"***\", \"**\", \"*\", \"B\", \"all\"],\n",
    "                                  name=\"Minimum z scoring for nodes:\", value=\"**\")\n",
    "input_sq_minrep = pn.widgets.Select(options=[1, 2, 3], name=\"Minimum shared replicates:\", value=2)\n",
    "input_sq_maxq = pn.widgets.Select(options=[\"1%\", \"5%\", \"10%\", \"25%\", \"50%\", \"100%\"],\n",
    "                                  name=\"Maximum quantile for edges:\", value=\"50%\")\n",
    "input_sq_highlight = pn.widgets.AutocompleteInput(options=[el for el in meta_dict[\"gene_id\"].keys() if type(el) == str]+\n",
    "                                                  [\"none\", \"None\"], name=\"Highlight gene:\", value=\"None\")\n",
    "input_sq_disablehvnx = pn.widgets.Checkbox(value=False, name=\"figure style\")\n",
    "\n",
    "# Options for barplot\n",
    "input_sq_bary = pn.widgets.Select(options=[\"Distance\", \"z-scoring based category\", \"Rank range between replicates\",\n",
    "                                           \"Percentile change F3 - full proteome\", \"Worst replicate correlation\",\n",
    "                                           \"Local distance percentile\"], name=\"Select y-axis to display:\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Functions for single query data display\n",
    "\n",
    "# Get the single query data\n",
    "def store_gene_data(event):\n",
    "    input_sq_button.disabled=True\n",
    "    output_sq_status.object = \"Calculating neighborhood ...\"\n",
    "    output = hv_query_df(df_core, meta_dict, gene=input_sq_gene.value, size=input_sq_topn.value_throttled,\n",
    "                         rank_tolerance=input_sq_tolerance.value_throttled, min_corr=input_sq_minr.value_throttled,\n",
    "                         min_enr=input_sq_minenr.value_throttled, perc_area=input_sq_nhsize.value_throttled)\n",
    "    if len(output) == 3:\n",
    "        (df, q, msg) = output\n",
    "        output_sq_status.object = \"Done calculating neighborhood\"\n",
    "        cache_sq_q.object = q.to_json()\n",
    "        cache_sq_neighborhood.object = df.to_json()\n",
    "    else:\n",
    "        output_sq_status.object = output\n",
    "        cache_sq_q.object = \"\"\n",
    "        cache_sq_neighborhood.object = \"\"\n",
    "    input_sq_button.disabled=False\n",
    "input_sq_button.on_click(store_gene_data)\n",
    "\n",
    "# Barplot\n",
    "@pn.depends(cache_sq_neighborhood.param.object, input_sq_bary.param.value)\n",
    "def get_gene_data(df, var):\n",
    "    try:\n",
    "        df = pd.read_json(df).sort_values(\"Distance measure\")\n",
    "    except:\n",
    "        return \"\"\n",
    "    p = sq_barplot(df, var)\n",
    "    return p\n",
    "\n",
    "# Networkplot\n",
    "@pn.depends(cache_sq_neighborhood.param.object,\n",
    "            input_sq_minz.param.value, input_sq_maxq.param.value, input_sq_minrep.param.value)\n",
    "def layout_single_network(query, min_z, max_q, min_rep):\n",
    "    # Retrieve neighborhood information\n",
    "    try:\n",
    "        query_result = pd.read_json(query).sort_values(\"Distance measure\")\n",
    "        q = pd.read_json(cache_sq_q.object, typ=\"series\", convert_dates=False, convert_axes=False)\n",
    "    except:\n",
    "        cache_sq_nwdists.object = \"\"\n",
    "        cache_sq_gp.object = \"\"\n",
    "        return \"Failed to read neighborhood (see settings panel for details).\"\n",
    "    \n",
    "    # Define node list\n",
    "    z_dict = {\"all\":5, \"B\":4, \"*\":3, \"**\":2, \"***\":1}\n",
    "    min_z = z_dict[min_z]\n",
    "    try:\n",
    "        nwk_members = filter_nwk_members(query_result, min_rep, min_z)\n",
    "    except:\n",
    "        cache_sq_nwdists.object = \"\"\n",
    "        cache_sq_gp.object = \"\"\n",
    "        return \"Failed to generate node list.\"\n",
    "    \n",
    "    # Layout network\n",
    "    q_dict = {\"1%\":0.01, \"5%\":0.05, \"10%\":0.1, \"25%\":0.25, \"50%\":0.5, \"100%\":1}\n",
    "    try:\n",
    "        dists_pd, GP = layout_network(df_core, query_result, nwk_members, q_dict[max_q], q, layout_method=\"Spring\")\n",
    "    except:\n",
    "        cache_sq_nwdists.object = \"\"\n",
    "        cache_sq_gp.object = \"\"\n",
    "        return \"Failed to layout network.\"\n",
    "    \n",
    "    # Store distances and layout in caching panels\n",
    "    cache_sq_nwdists.object = dists_pd.to_json()\n",
    "    cache_sq_gp.object = pd.DataFrame(GP).to_json()\n",
    "    \n",
    "    return \"\"\n",
    "\n",
    "@pn.depends(cache_sq_gp.param.object, input_sq_highlight.param.value, input_sq_disablehvnx.param.value)\n",
    "def draw_single_network(GP, highlight, figure_style):\n",
    "    # Read query result and quantile cutoffs\n",
    "    try:\n",
    "        query_result = pd.read_json(cache_sq_neighborhood.object).sort_values(\"Distance measure\")\n",
    "        q = pd.read_json(cache_sq_q.object, typ=\"series\", convert_dates=False, convert_axes=False)\n",
    "    except:\n",
    "        return \"\"\n",
    "    \n",
    "    # Read distances and network layout from cache\n",
    "    try:\n",
    "        dists_pd = pd.read_json(cache_sq_nwdists.object)\n",
    "        GP = pd.read_json(GP).to_dict()\n",
    "        GP = {k: np.array(list(GP[k].values())) for k in GP.keys()}\n",
    "    except:\n",
    "        return \"\"\n",
    "    \n",
    "    # switch between figure and interactive style\n",
    "    if figure_style:\n",
    "        nwk = draw_network_figure(dists_pd, GP, query_result, highlight, q)\n",
    "    else:\n",
    "        nwk = draw_network_interactive(dists_pd, GP, query_result, highlight, q)\n",
    "    \n",
    "    return nwk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Assemble output tabs for single query\n",
    "output_bar = pn.Column(input_sq_bary, pn.pane.Markdown(\"colors indicate occurence across replicate lists\"), get_gene_data)\n",
    "output_nwk = pn.Column(pn.Row(input_sq_minz, input_sq_maxq, input_sq_minrep),\n",
    "                       pn.Row(input_sq_highlight, input_sq_disablehvnx),\n",
    "                       pn.Row(draw_single_network, pn.pane.SVG(\"./LegendNetworksFull.svg\")),\n",
    "                       layout_single_network)\n",
    "\n",
    "output_sq_tabs = pn.Tabs()\n",
    "output_sq_tabs.append((\"Network plot\", output_nwk))\n",
    "output_sq_tabs.append((\"Barplot\", output_bar))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Interface elements for multiple network query\n",
    "\n",
    "# Caching panels for multiqueries\n",
    "cache_mq_query = pn.pane.Str(\"\")\n",
    "cache_mq_q = pn.pane.Str(\"\")\n",
    "cache_mq_nwdists = pn.pane.Str(\"\")\n",
    "cache_mq_gp = pn.pane.Str(\"\")\n",
    "cache_mq_settings = pn.pane.Str(\"\")\n",
    "\n",
    "# Main options\n",
    "input_mq_preselected = pn.widgets.Select(options=[\"CD63, CD81, CD3G\", \"Tetraspanins\", \"EV Markers from PCA\"],\n",
    "                                         name=\"Predefined \\n gene lists\")\n",
    "input_mq_list = pn.widgets.TextAreaInput(value=\"CD63, CD81, CD3G\", sizing_mode=\"fixed\")\n",
    "input_mq_topn = pn.widgets.IntSlider(start=20, end=100, step=5, value=30, value_throttled=30,\n",
    "                                     name=\"Neighborhood size\")\n",
    "input_mq_button = pn.widgets.Button(name='Calculate neighborhoods', button_type='primary')\n",
    "output_mq_status = pn.pane.Markdown(\"\", width_policy=\"max\", sizing_mode=\"stretch_width\")\n",
    "\n",
    "# Advanced options\n",
    "input_mq_tolerance = pn.widgets.IntSlider(start=5, end=50, step=5, value=45, value_throttled=45,\n",
    "                                          name=\"Replicate neighborhood tolerance\")\n",
    "input_mq_minr = pn.widgets.FloatSlider(start=-1, end=1, step=0.05, value=0, value_throttled=0,\n",
    "                                       name=\"Minimal profile correlation\")\n",
    "input_mq_minenr = pn.widgets.FloatSlider(start=-1, end=1, step=0.05, value=-1, value_throttled=-1,\n",
    "                                         name=\"Minimal percentile shift in F3 vs. full proteome\")\n",
    "input_mq_nhsize = pn.widgets.IntSlider(start=50, end=500, step=50, value=250, value_throttled=250,\n",
    "                                       name=\"Neighborhood size for scoring\")\n",
    "\n",
    "# Tabcolumn containing options\n",
    "options_multi = pn.Tabs()\n",
    "options_multi.append((\"Neighborhood selection\", pn.Column(input_mq_preselected, \n",
    "                                                          pn.pane.Markdown(\"List of gene symbols to compare  \\n\\\n",
    "                                                          (select from above for examples)\"),\n",
    "                                                          input_mq_list, input_mq_topn,\n",
    "                                                          input_mq_button, output_mq_status)))\n",
    "options_multi.append((\"Advanced options\", pn.Column(input_mq_tolerance, input_mq_minr,\n",
    "                                                    input_mq_minenr, input_mq_nhsize)))\n",
    "\n",
    "# Options for multi query network\n",
    "input_mq_minz = pn.widgets.Select(options=[\"***\", \"**\", \"*\", \"B\", \"all\"],\n",
    "                                  name=\"Minimum z scoring for nodes:\", value=\"B\")\n",
    "input_mq_minrep = pn.widgets.Select(options=[1, 2, 3], name=\"Minimum shared replicates:\", value=2)\n",
    "input_mq_maxq = pn.widgets.Select(options=[\"1%\", \"5%\", \"10%\", \"25%\", \"50%\", \"100%\"],\n",
    "                                  name=\"Maximum quantile for edges:\", value=\"50%\")\n",
    "input_mq_highlight = pn.widgets.AutocompleteInput(options=[el for el in meta_dict[\"gene_id\"].keys() if type(el) == str]+\n",
    "                                                  [\"none\", \"None\"], name=\"Highlight gene:\", value=\"None\")\n",
    "input_mq_disablehvnx = pn.widgets.Checkbox(value=False, name=\"figure style\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Functions for multi query data display\n",
    "\n",
    "# Put in predefined gene sets\n",
    "def update_genelist(event):\n",
    "    l = event.new\n",
    "    lg_dict = {\"CD63, CD81, CD3G\": \"CD63, CD81, CD3G\",\n",
    "               \"Tetraspanins\": \"CD81, CD9, CD63, CD151, TSPAN33, CD82, CD53, TSPAN5, TSPAN7\",\n",
    "               \"EV Markers from PCA\": \"TTYH3, PLSCR3, MINK1, SELPLG, IGSF8, CD9, SLC44A1, CD53, CD82, TNFAIP3, \\\n",
    "                ARRDC1, ALCAM, CD63, TAOK1, MAP4K4, VPS28, PPP1R18, CD81, CHMP2A, PIP4K2A, PDCD6, SH3GL1, VPS4B, \\\n",
    "                CHMP2B, CHMP1A, VPS4A, ANXA7, GDI2, TRIM25, LDHA, KARS\"}\n",
    "    if l in lg_dict.keys():\n",
    "        input_mq_list.value = lg_dict[l]\n",
    "input_mq_preselected.param.watch(update_genelist, 'value')\n",
    "    \n",
    "\n",
    "# Get the multi query data -> change this to a button triggered callback and create a second callback to\n",
    "# validate the input, compare the settings and enable/disable the run button accordingly\n",
    "# @pn.depends(input_mq_list.param.value, input_mq_topn.param.value_throttled,\n",
    "#             input_mq_tolerance.param.value_throttled, input_mq_minr.param.value_throttled,\n",
    "#             input_mq_minenr.param.value_throttled, input_mq_nhsize.param.value_throttled)\n",
    "@pn.depends(input_mq_list.param.value, input_mq_topn.param.value_throttled,\n",
    "            input_mq_tolerance.param.value_throttled, input_mq_minr.param.value_throttled,\n",
    "            input_mq_minenr.param.value_throttled, input_mq_nhsize.param.value_throttled)\n",
    "def validate_mq_param(genes, size, tolerance, minr, minenr, nhsize):\n",
    "    wdgts = [input_mq_preselected, input_mq_list, input_mq_topn, input_mq_tolerance,\n",
    "             input_mq_minr, input_mq_minenr, input_mq_nhsize]\n",
    "    for wdgt in wdgts:\n",
    "        wdgt.disabled = True\n",
    "    genes = [el.strip().upper() for el in input_mq_list.value.split(\",\")]\n",
    "    genes.sort()\n",
    "    for el in genes:\n",
    "        if el not in meta_dict['gene_id'].keys():\n",
    "            output_mq_status.object = \"{} not found in the dataset. The single gene query tab has \\\n",
    "                                       an autocomplete function for all quantified genes. Use this to check, \\\n",
    "                                       which genes can be put into the mutli gene query.\".format(el)\n",
    "            for wdgt in wdgts:\n",
    "                wdgt.disabled = False\n",
    "            input_mq_button.disabled = True\n",
    "            return\n",
    "    oldparams = cache_mq_settings.object\n",
    "    newparams = {\"gene\": genes, \"size\": size, \"tolerance\": tolerance, \"minr\": minr, \"minenr\": minenr, \"nhsize\": nhsize}\n",
    "    newparams = pd.DataFrame(pd.Series(newparams)).to_json()\n",
    "    if oldparams == newparams:\n",
    "        output_mq_status.object = \"No need to recalculate, the result for these settings is what you are looking at.\"\n",
    "        for wdgt in wdgts:\n",
    "            wdgt.disabled = False\n",
    "        input_mq_button.disabled = True\n",
    "        return\n",
    "    else:\n",
    "        output_mq_status.object = \"Hit the button to start calculating the selected neighborhoods. \\\n",
    "        This can take ~0.5 min per gene.\"\n",
    "        for wdgt in wdgts:\n",
    "            wdgt.disabled = False\n",
    "        input_mq_button.disabled = False\n",
    "        return\n",
    "\n",
    "\n",
    "def store_mq_data(event):\n",
    "    wdgts = [input_mq_preselected, input_mq_list, input_mq_topn, input_mq_tolerance,\n",
    "             input_mq_minr, input_mq_minenr, input_mq_nhsize]\n",
    "    for wdgt in wdgts:\n",
    "        wdgt.disabled = True\n",
    "    input_mq_button.disabled = True\n",
    "    output_mq_status.object = \"Started neighborhood calculation ...\"\n",
    "    genes = [el.strip().upper() for el in input_mq_list.value.split(\",\")]\n",
    "    genes.sort()\n",
    "    params = {\"gene\": genes, \"size\": input_mq_topn.value_throttled, \"tolerance\": input_mq_tolerance.value_throttled,\n",
    "              \"minr\": input_mq_minr.value_throttled, \"minenr\": input_mq_minenr.value_throttled,\n",
    "              \"nhsize\": input_mq_nhsize.value_throttled}\n",
    "    cache_mq_settings.object = pd.DataFrame(pd.Series(params)).to_json()\n",
    "    pb = pn.widgets.Progress(max=len(genes), value=0, width_policy=\"max\")\n",
    "    options_multi[0].append(pb)\n",
    "    output = multi_query(df_core, meta_dict, genes=genes,\n",
    "                         size=params[\"size\"], rank_tolerance=params[\"tolerance\"], min_corr=params[\"minr\"],\n",
    "                         min_enr=params[\"minenr\"], perc_area=params[\"nhsize\"], pb=pb)\n",
    "    options_multi[0].remove(pb)\n",
    "    if len(output) == 3:\n",
    "        (df, q, msg) = output\n",
    "        output_mq_status.object = \"Done calculating neighborhoods.\"\n",
    "        cache_mq_q.object = q.to_json()\n",
    "        cache_mq_query.object = df.to_json()\n",
    "    else:\n",
    "        output_mq_status.object = output\n",
    "        cache_mq_q.object = \"\"\n",
    "        cache_mq_query.object = \"\"\n",
    "    for wdgt in wdgts:\n",
    "        wdgt.disabled = False\n",
    "input_mq_button.on_click(store_mq_data)\n",
    "\n",
    "\n",
    "# Networkplot\n",
    "@pn.depends(cache_mq_query.param.object, input_mq_minz.param.value, input_mq_maxq.param.value, input_mq_minrep.param.value)\n",
    "def layout_multi_network(query, min_z, max_q, min_rep):\n",
    "    # Retrieve neighborhood information\n",
    "    try:\n",
    "        query_result = pd.read_json(query).sort_values(\"Distance measure\")\n",
    "        q = pd.read_json(cache_mq_q.object, typ=\"series\", convert_dates=False, convert_axes=False)\n",
    "    except:\n",
    "        cache_mq_nwdists.object = \"\"\n",
    "        cache_mq_gp.object = \"\"\n",
    "        return \"Failed to read neighborhood (see settings panel for details).\"\n",
    "    \n",
    "    # Define node list\n",
    "    z_dict = {\"all\":5, \"B\":4, \"*\":3, \"**\":2, \"***\":1}\n",
    "    min_z = z_dict[min_z]\n",
    "    try:\n",
    "        nwk_members = filter_nwk_members(query_result, min_rep, min_z)\n",
    "    except:\n",
    "        cache_mq_nwdists.object = \"\"\n",
    "        cache_mq_gp.object = \"\"\n",
    "        return \"Failed to generate node list.\"\n",
    "    \n",
    "    # Layout network\n",
    "    q_dict = {\"1%\":0.01, \"5%\":0.05, \"10%\":0.1, \"25%\":0.25, \"50%\":0.5, \"100%\":1}\n",
    "    try:\n",
    "        dists_pd, GP = layout_network(df_core, query_result, nwk_members, q_dict[max_q], q, layout_method=\"Spring\")\n",
    "    except:\n",
    "        cache_mq_nwdists.object = \"\"\n",
    "        cache_mq_gp.object = \"\"\n",
    "        return \"Failed to layout network.\"\n",
    "    \n",
    "    # Store distances and layout in caching panels\n",
    "    cache_mq_nwdists.object = dists_pd.to_json()\n",
    "    cache_mq_gp.object = pd.DataFrame(GP).to_json()\n",
    "    \n",
    "    return \"\"\n",
    "\n",
    "@pn.depends(cache_mq_gp.param.object, input_mq_highlight.param.value, input_mq_disablehvnx.param.value)\n",
    "def draw_multi_network(GP, highlight, figure_style):\n",
    "    # Read query result and quantile cutoffs\n",
    "    try:\n",
    "        query_result = pd.read_json(cache_mq_query.object).sort_values(\"Distance measure\")\n",
    "        q = pd.read_json(cache_mq_q.object, typ=\"series\", convert_dates=False, convert_axes=False)\n",
    "    except:\n",
    "        return \"\"\n",
    "    \n",
    "    # Read distances and network layout from cache\n",
    "    try:\n",
    "        dists_pd = pd.read_json(cache_mq_nwdists.object)\n",
    "        GP = pd.read_json(GP).to_dict()\n",
    "        GP = {k: np.array(list(GP[k].values())) for k in GP.keys()}\n",
    "    except:\n",
    "        return \"\"\n",
    "    \n",
    "    # switch between figure and interactive style\n",
    "    if figure_style:\n",
    "        nwk = draw_network_figure(dists_pd, GP, query_result, highlight, q)\n",
    "    else:\n",
    "        nwk = draw_network_interactive(dists_pd, GP, query_result, highlight, q)\n",
    "    \n",
    "    return nwk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Assemble output tabs for multi query\n",
    "output_mq_nwk = pn.Column(pn.Row(input_mq_minz, input_mq_maxq, input_mq_minrep),\n",
    "                          pn.Row(input_mq_highlight, input_mq_disablehvnx),\n",
    "                          pn.Row(draw_multi_network, pn.pane.SVG(\"./LegendNetworksFull.svg\")),\n",
    "                          layout_multi_network)\n",
    "\n",
    "output_mq_tabs = pn.Tabs()\n",
    "output_mq_tabs.append((\"Network plot\", output_mq_nwk))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Aboutstab\n",
    "about = '''\n",
    "## Quick Guide:\n",
    "1. Enter a gene symbol in UPPER CASE in the single query tab or several in the multi query tab.\n",
    "2. Start the neighborhood calculation by clicking the button underneath.\n",
    "3. The network shows the distance based relationships between the close neighbours, ideally\n",
    "revealing tight clusters (try e.g. PSMA1) and peripheral proteins.\n",
    "4. The barplot shows various quantitative measures for all neighbors (only in single query mode).\n",
    "5. In order to understand what the advanced settings and additional parameters for the network\n",
    "selection, please refer to the theoretical section below. All settings have reasonable default values.\n",
    "\n",
    "### Useful hints\n",
    "- The checkbox underneath the settings for the network layout lets you switch between the interactive network, \n",
    "which includes a toolbar to the right as provided by hvplot, and the figure style (7pt font-size) network, \n",
    "which can be copied for future reference (Download option pending).\n",
    "- If network nodes are too crowded, reduce the stringency on the quantile to add back more edges that pull \n",
    "distant nodes closer together and thereby tight clusters slightly apart.\n",
    "- The inclusion of neighbours found in only one neighbourhood replicate is not recommended, but can be tried \n",
    "for an low stringency explorative analysis of the data. \n",
    "- If you expect a protein to appear in the neighborhood of a query but can’t see it in the network, try \n",
    "increasing the neighborhood size and tolerance, relief the z-score stringency and use the highlight option \n",
    "on the network to find the protein.\n",
    "- The single gene input autocompletes. If a protein is not found, check on Uniprot if it has a \n",
    "different primary gene name. If it is still not on the list it was not quantified sufficiently.\n",
    "\n",
    "## Theoretical framework and interpretation of results\n",
    "\n",
    "### Neighborhood calculation\n",
    "Profile similarity between proteins is evaluated by calculating the pairwise \n",
    "Manhattan distance (absolute summed difference at every point of the two profiles). The query  \n",
    "is retrieved as the top hit (with a distance of 0), and other proteins (see single query barplot) are \n",
    "shown in order of increasing distance to the query. The neighbourhood size defines how many proteins \n",
    "will be displayed in the barplot, and used for building a neighbourhood network. In the advanced \n",
    "settings quality filters can be applied to adjust the stringency and sensitivity of the analysis:\n",
    "- minimum number of SILAC quantification events (to be added)\n",
    "- minimum cosine correlation of profiles across repeats\n",
    "- abundance percentile shift in the F3 extracellular vesicle fraction, which contains the smallest vesicles \n",
    "in the highest proportion, relative to the full proteome\n",
    "Proteins not matching these criteria are removed from the dataset before any distances are calculated.\n",
    "\n",
    "### Neighborhood evaluation\n",
    "There are three measures provided to gauge which proteins are reproducibly in close proximity to the \n",
    "query, in the context of the local data density.\n",
    "\n",
    "1. Replicate ranking  \n",
    "The barplot shows the difference between the best and the worst proximity rank of a protein \n",
    "across the three replicates. If the top scoring neighbor (most likely rank 1 in at least one of the \n",
    "replicates) has a replicate range of 10 this means that is was at least on rank 11 in one of the \n",
    "other replicates. A protein is considered common to all three replicates if its overall rank is within \n",
    "the specified neighborhood size and each replicate rank is within the set neighbourhood size + ‘tolerance’, \n",
    "which can be set in the advanced settings.\n",
    "2. Z-scoring of proximity  \n",
    "To define the close protein neighbourhood of a query, in the context of the local dataspace, \n",
    "a simple estimate of the local profile density and distribution is performed. From the nearest 250 proteins \n",
    "(neighborhood size for scoring as set in the advanced settings), the median distance to the query \n",
    "(termed MDQ) is calculated. Next, for every one of these 250 proteins the absolute distance to the MDQ is \n",
    "determined. The median of these values corresponds to the MAD (median absolute distance to the MDQ). Each \n",
    "positive distance to the MDQ (ie the right tail of the distribution) is then transformed to a pseudo Z \n",
    "score, using a robust estimate of the standard deviation: pZ = ((distance to MDQ) – MDQ) / (1.483 x MAD). \n",
    "Profiles are categorized by these pZ scores: very close neighbours (Z>3.09, \\*\\*\\*), close (Z>2.33, \\*\\*), \n",
    "fairly close (Z>1.64, *), borderline (Z>1.28, B). Note that these categories provide \n",
    "guidance to evaluate relative profile proximity in the context of other mapped proteins nearby. However, they \n",
    "neither reflect exact boundaries for local clusters, nor exact probabilities of association with the query \n",
    "protein. While the actual distances between two proteins are identical regardless \n",
    "which is submitted as the query, the proximity guide classifier may be different because the set \n",
    "of proteins defining the distribution changes with the query. For example, an isolated protein near the \n",
    "edge of a dense cluster of profiles will retrieve some of these proteins as relatively close neighbours, but \n",
    "not vice versa. Thus, it may also be informative to consider also the absolute distances of proteins to the \n",
    "query, as indicated in the predictor output, and the distance quantile.\n",
    "3. Distance quantiles  \n",
    "To achieve an additional non-directional classification of distances, which is essential for network \n",
    "construction, all pairwise Manhattan-distances between the query’s 250 closest neighbours (same setting \n",
    "as for z-scoring) were calculated. The distances between the network nodes are then categorized by the \n",
    "quantile they occupy in this distribution. If a neighbor has a distance quantile of <1% it means it \n",
    "is at least as close to the query as the closest 1% of all protein pairs in that area are to each other (i.e. \n",
    "a very close neighbor). This score is non-directional, but is affected by other data structures in the \n",
    "vicinity. Again considering an isolated protein close to a protein cluster, the nearest neighbours will \n",
    "occupy a higher quantile because the pairwise distances within the protein cluster will shift the overall \n",
    "distribution towards smaller distances.\n",
    "\n",
    "### Network generation\n",
    "To visualize the complex structure of local neighbourhoods, we utilized network analysis. This can reveal \n",
    "for example if the closest neighbours of a protein form a very tight network among themselves, indicating \n",
    "a functional cluster; if the neighbourhood contains one or more tight subclusters, which may correspond to \n",
    "protein complexes; or if the local neighbourhood is very evenly distributed. These insights cannot be judged \n",
    "from a PCA plot alone, as the dimensional flattening loses information contained in the original \n",
    "nine-dimensional dataset. Similarly, the network architecture is not apparent from the linear list of nearest \n",
    "neighbours (the bar plot), as the relative proximity of the neighbours is not considered; hence, two proteins \n",
    "that are both close to the query might be quite distant from each other. The network analysis was carried \n",
    "out in python using the networkx library \n",
    "[(Hagberg et al, 2008)](https://networkx.github.io/documentation/stable/index.html). Network nodes are \n",
    "selected based on the filter criteria described above, from the neighborhood as calculated based on the \n",
    "selected settings. Nodes have to be close neighbours (by rank) in at least a defined number of replicates \n",
    "(nodes are colour coded accordingly), and have to achieve a minimal z-score, as set by the user. The distances \n",
    "between the remaining network nodes are then categorized by the quantile scoring, calculated by scoring all \n",
    "pairwise connections within this set (not just the ones to the query). All edges that fall into a higher \n",
    "quantile than the selected cut off are discarded. If a node is completely disconnected by this it will not be \n",
    "displayed in the network. All remaining distances are inverted and used as edgeweights for a force-directed \n",
    "layouting algorithm (spring network layout (Fruchterman & Reingold, 1991)), which pulls proteins with short \n",
    "distances closer together. Edges are colored by their distance percentile to visually reveal tight clusters.\n",
    "\n",
    "### Multi query networks\n",
    "To visualize the segregation/overlap of adjacent neighbourhoods, networks can be constructed from more \n",
    "than one query protein. For each of the individual queries, nodes are selected as above. Here a slight \n",
    "adjustment of parameters is useful: Fewer neighbors should be considered, to clearly separate distant queries, \n",
    "but borderline distances should be included, to allow for more connections between adjacent neighbourhoods to \n",
    "shape the network layout. The quantile boundaries are calculated for each individual set of 250 neighbors and \n",
    "then averaged across the queries. Thus, if two queries from areas of dataspace with different densities are \n",
    "selected, this will become apparent in the number and color of the edges (the sparser neighbourhood will have \n",
    "fewer/thinner connections).\n",
    "\n",
    "## Implementation notes\n",
    "This site was implemented in python using the holoviz framework and its bindings to several other python \n",
    "libraries (<https://holoviz.org>). The source code for this website and the corresponding analysis is stored \n",
    "at <https://github.com/JuliaS92/EVProfiler>.\n",
    "\n",
    "## Reference\n",
    "Please alsways include the original research publication when referencing this site:\n",
    "\n",
    "Lorena Martin-Jaular, Nathalie Nevo, Julia P. Schessner, Mercedes Tkach, Mabel Jouve, Florent Dingli, Damarys Loew, \n",
    "Kenneth  W. Witwer,  Matias Ostrowski, Georg H.H. Borner, Clotilde Théry. Proteomic profiling allows unbiased analysis \n",
    "of HIV-1 and host extracellular vesicles. EMBO Journal (2021).\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "## Assemble full app\n",
    "content = pn.Tabs()\n",
    "content.append((\"Single gene query\", pn.Row(pn.Column(options_single), output_sq_tabs)))\n",
    "content.append((\"Multi gene query\", pn.Row(pn.Column(options_multi, validate_mq_param), output_mq_tabs)))\n",
    "content.append((\"About\", pn.pane.Markdown(about)))\n",
    "\n",
    "app_container = pn.Column(\"# Jurkat EV Neighbour Network Predictor Tool\", content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "app_container.servable()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
